changefeedccl: Flush aggregators if previous flush was skipped

miretskiy opened the pull request:
Previously, if change aggregator skipped progress update flush due
to flush happening too frequently, the next time flush would happen
would be only during next frontier update.

Next frontier update could take a while, particularly when
dealing with large tables.  Ensure that if the flush was skipped
before, then as soon as flush frequency falls below the threshold,
issue previously skipped flush even if frontier did not advance.

Release justification: low danger change to improve flush behavior.
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86500)
<!-- Reviewable:end -->

samiskin commented on pkg/ccl/changefeedccl/changefeed_processors.go:
Is there a benefit to changing `lastFlush` to mean "the start of the last flush attempt" as opposed to "the last completion of a flush"?

miretskiy replied:
Ohh... That's what it was :)
Just didn't want to keep defer.... I can bring it back.

miretskiy replied:
Reverted to using defer.

samiskin reviewed (commented):
If we had a high traffic table where every event advanced the frontier a tiny bit, it looks like now we'd begin Flushing every other resolved event.

Its a little scary to me that in this high traffic case we'd now be increasing the number of messages we send between nodes, as well as increasing the amount of blocking operations we do where we send some messages, block on flush, and then allow more messages to be sent.  It feels like we'd be decreasing our throughput specifically in these high traffic cases.

Does that sound like a legitimate concern?  I'm not sure just how often we'd get resolved spans within the aggregators.

Commit:
changefeedccl: Ensure correct ordering.

When async flushing enabled, the following sequence
of events is possible (even if unlikely):
1. `k@t1` is emitted, causing async flush to write file `f1`
2. `k@t2` is emitted, causing async flush to write file `f2`
3. `f2` is written out before `f1`.

In this unlikely scenario -- and the reason why it's unlikely is that
we have to generate megabytes of data to cause a flush, unless,
file_size parameter was pathologically small -- if a client
were to read the contents of the directory right after step 2,
and then read directory after step 3, the client will first observe
`k@t2`, and then observe `k@t1` -- which is a violation of
ordering guarantees.

A simple fix to the above problem is to simply maintain
the queue of pending flushes.  This approach is simple, it still
has a reasonalbe throughput since Emit rows are not blocked.
However it results in somewhat lower overall throughput since
there is only 1 file being written out at any time; thus,
the queue tends to build up, and pushback mechanism slows down the
throughput.

Instead, this PR adopts an approach to allow throughput sensitive
phases -- i.e. initial scan and schema change backfills -- to happen
fully asynchronously, while ensuring that during normal operations,
file reordering, as described above, cannot happen.

Release note: None.

shermanCRL commented:
Do we intend to do anything with this?

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=86500) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=86500) before we can accept your contribution.<br/><hr/>**Yevgeniy Miretskiy** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=86500) it.</sub>
